Cudnn deconv

wangyang59 opened the pull request:
Implemented cudnn_convt, convTransProjection and convTransOperator layers.

wangkuiyi commented:
@wangyang59 I think that it's because  @qingqing01 has been working hard on the new API and related issues, so she's been a little slowly on reviewing this PR.

wangyang59 commented:
@wangkuiyi 
Sure, no problem. I totally understand.

qingqing01 commented on paddle/gserver/layers/CudnnConvTransLayer.cpp:
```diff
@@ -0,0 +1,23 @@
+
+namespace paddle {
+
+REGISTER_LAYER(cudnn_convt, CudnnConvTransLayer);
```
文件较多，继承在这里并没起啥作用：

```
class CudnnConvTransLayer : public CudnnConvBaseLayer {
class CudnnConvLayer : public CudnnConvBaseLayer {
```

父类``` CudnnConvBaseLayer ```实现的所有的东西，包含ConvTrans和Conv的功能，而两个子类几乎没有区别，除了REGISTER_LAYER()时的类型。

这样继承较多，文件较多，其实有点晕~

qingqing01 replied:
我在想 ConvProjection 和 ConvTransProjection已经区分了conv和conv trans的实现了，CudnnConvTransLayer， CudnnConvLayer，CudnnConvBaseLayer可以合并成一个？ 只在python的用户接口里区分conv 和 conv trans？或许可以在想想更合理的结构，现在这样觉得有点复杂

wangyang59 commented:
@qingqing01 Thanks a lot for the comments. 
I have consolidate the 4 files (CudnnConvLayer.h/cpp, CudnnConvTransLayer.h/cpp) into 1 file (CudnnConvLayer.cpp) to make things simpler. 

If you would like to remove them completely, I can add a `isDeconv` flag in the `ConvConfig` proto which might be confusing to users since conv/deconv can be set both in isDeconv flag and layer type.

wangyang59 commented:
@qingqing01 
I have changed the code according to our discussion yesterday. I added two REGISTER_LAYER function calls to CudnnConvBaseLayer, and removed CudnnConvLayer.cpp.

qingqing01 commented on paddle/gserver/layers/ConvBaseProjection.cpp:
```diff
@@ -0,0 +1,207 @@
+        << "Wrong input size for convolution"
+        << " channels=" << configChannels_ << " imageH=" << imageH_
+        << " imageW=" << imageW_ << " inputSize=" << in_->value->getWidth();
+  }
```
configChannels_  -> channels_

另外，check如果不是必须的就不要了，减少在父类里增加这样if else分支支持不同子类的功能。

qingqing01 commented on python/paddle/trainer_config_helpers/layers.py:
```diff
@@ -703,8 +703,10 @@ def __enter__(self):
+    def __exit__(self, exc_type, exc_value, tb):
+        if exc_type != None:
+            traceback.print_exception(exc_type, exc_value, tb)
+            sys.exit(1)
```
这里为啥要改？

wangyang59 replied:
I changed here because I found that with the original setting, if there is something wrong when defining the mixed layer (e.g., some error in one of the projection layer), then the error message would not reflect the position/reason of the actual error. So I decided to change it so that the relevant error messages will be displayed

reyoung replied:
```python
if exc_value is not None:
    raise exc_value
```

qingqing01 commented on paddle/gserver/layers/ConvBaseOperator.cpp:
```diff
@@ -0,0 +1,197 @@
+    CHECK_EQ(outputH_, outputY_);
+    CHECK_EQ(outputW_, outputX_);
+    out_->setFrameHeight(outputH_);
+    out_->setFrameWidth(outputW_);
```
if (isDeconv_) {} else {}分支，可以放到子类里实现，ConvProjection同理。

qingqing01 commented on paddle/gserver/layers/ConvBaseProjection.h:
```diff
@@ -0,0 +1,162 @@
+      inputOffset_ = (configChannels_ / groups_) * imageH_ * imageW_;
+      outputOffset_ = (configNumFilters_ / groups_) * outputH_ * outputW_;
+      return outputH_ * outputW_ * configNumFilters_;
+    }
```
calOutputSize放到子类里去实现。

wangyang59 commented:
@qingqing01 
Thanks a lot for your comments. I have modified the codes according to your suggestions by putting deconv dependent parts into derived classes.

wangyang59 commented:
@reyoung @qingqing01 
Thanks for your review. I have changed the codes accordingly. Please help merge the PR if there are no further issues.
